File: bindings/node/perf/multi/perf_multi_router_router_server.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const readline = require('node:readline');
const zlink = require('@zlink-systems/zlink');
const { configureTlsServer } = require('../common/perf_tls');
const { parseMultiArgs } = require('./perf_multi_common');
const {
  POLLIN,
  POLLOUT,
  applyAutoHwmMsgUnit,
  applyContextPolicy,
  applySocketPolicy,
  emitMultiSocketHwmDetail,
  pollEvents,
  pollEventHas,
  trySocketSend,
  waitPollerOne,
  waitForConnectionReadyCount
} = require('./perf_multi_runtime');

function drainPending(router, pending) {
  while (pending.length > 0) {
    const reply = pending[0];
    if (!trySocketSend(router, reply.routingId, reply.payload)) {
      break;
    }
    pending.shift();
  }
}

function receiveAndQueueReplies(router, pending) {
  while (true) {
    const received = new zlink.Received();
    if (!router.recv(received, zlink.RecvFlags.DontWait)) {
      break;
    }
    if (!received.routingId || received.requestSeq) {
      continue;
    }
    const routingId = received.routingId;
    const payload = received.singlePartOrThrow();
    if (pending.length === 0 && trySocketSend(router, routingId, payload)) {
      continue;
    }
    pending.push({ routingId, payload: Buffer.from(payload.data()) });
  }
}

async function main() {
  const options = parseMultiArgs(process.argv.slice(2));
  const ctx = zlink.createContext();
  applyContextPolicy(ctx, 'server', 'MULTI_ROUTER_ROUTER');
  const router = zlink.createRouterSocket(ctx);
  const poller = zlink.createPoller();
  const pending = [];
  let pollBuffer = null;
  let rl = null;
  let stop = false;

  try {
    applySocketPolicy(router);
    configureTlsServer(router, options.transport);
    router.setRoutingId(
      zlink.RoutingId.from(Buffer.from('multi-router-router-server', 'ascii'))
    );
    router.bind(options.endpoint);
    applyAutoHwmMsgUnit(ctx, options.msgSize);
    ctx.recalculateAutoHwm();
    emitMultiSocketHwmDetail(router, 'endpoint', options.transport, options.msgSize);
    poller.add(router, pollEvents(POLLIN), 0);
    pollBuffer = zlink.createPollEvents(1);
    const readyBarrier = waitForConnectionReadyCount(router, options.clients);
    console.log(`READY,${options.endpoint}`);

    rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
    (async () => {
      for await (const line of rl) {
        if (line === 'STOP' || line === 'QUIT') {
          stop = true;
          break;
        }
      }
    })();

    await readyBarrier;

    while (!stop) {
      poller.modify(router, pollEvents(POLLIN | POLLOUT));
      const ready = waitPollerOne(poller, pollBuffer, -1);
      if (!ready) {
        continue;
      }
      if (pollEventHas(ready, POLLOUT)) {
        drainPending(router, pending);
      }
      if (pollEventHas(ready, POLLIN)) {
        receiveAndQueueReplies(router, pending);
        drainPending(router, pending);
      }
    }
  } finally {
    rl?.close();
    pollBuffer?.close();
    poller.close();
    router.close();
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: bindings/python/perf/multi/perf_multi_router_router_server.py
import asyncio
import sys
import threading

import zlink

from perf_multi_common import (
    apply_multi_socket_options,
    benchmark_endpoint,
    configure_multi_tls_server,
    measurement_part_count,
    parse_server_args,
    perf_server_context,
    recv_nonblocking,
    RelaySchedulerQuantum,
    RoutedReplySender,
    safe_poll,
    scoped_relay_eager_task_factory,
)


async def main(argv=None):
    args = parse_server_args(argv or sys.argv[1:])
    endpoint = benchmark_endpoint(args.transport, "multi-router-router")
    stop = threading.Event()
    scheduler_quantum = RelaySchedulerQuantum()

    def wait_stop():
        for line in sys.stdin:
            if line.strip().upper() in {"STOP", "QUIT"}:
                stop.set()
                return
        # stdin EOF (parent closed pipe) is also a STOP signal.
        stop.set()

    threading.Thread(target=wait_stop, daemon=True).start()

    with perf_server_context() as ctx:
        with zlink.create_router_socket(ctx) as router:
            configure_multi_tls_server(router, args.transport)
            apply_multi_socket_options(router)
            router.set_routing_id(b"SERVER")
            router.bind(endpoint)
            print(f"READY,{endpoint}", flush=True)
            with zlink.create_poller() as poller:
                poller.add_socket(router, zlink.PollEventFlag.POLLIN, 0)
                poll_events = zlink.create_poll_events(1)
                recv_storage = zlink.create_received()
                replies = RoutedReplySender(router)
                # Python 3.12 starts each reply task through its public
                # coroutine immediately. Older runtimes retain create_task
                # scheduling and receive a bounded turn every quantum.
                with scoped_relay_eager_task_factory():
                    while not stop.is_set():
                        replies.raise_if_failed()
                        # The application FIFO may keep receiving snapshots,
                        # but its single sender awaits the preceding Core
                        # admission before submitting the next reply.
                        ready_count = safe_poll(poller, poll_events, 0)
                        for offset in range(ready_count):
                            if poll_events.slot(offset) != 0 or not (
                                poll_events.revents(offset)
                                & int(zlink.PollEventFlag.POLLIN)
                            ):
                                continue
                            while True:
                                received = recv_nonblocking(router, storage=recv_storage)
                                if received is None:
                                    break
                                with received:
                                    if len(received.parts) != measurement_part_count():
                                        raise RuntimeError("invalid measured multipart request")
                                    if len(received.parts) == 2 and len(received.parts[1].data) != 0:
                                        raise RuntimeError("invalid measured multipart trailing frame")
                                    payload = bytes(received.parts[0].data)
                                    routing_id = bytes(received.routing_id)
                                replies.enqueue(payload, routing_id)
                                # This is a scheduler fairness budget only. It
                                # neither caps pending replies nor gates Core
                                # admission on their count.
                                if scheduler_quantum.received():
                                    await asyncio.sleep(0)
                        scheduler_quantum.reset()
                        await asyncio.sleep(0)
                await replies.drain()


if __name__ == "__main__":
    asyncio.run(main())
